rbd-target-api.py

```diff
@@ -582,35 +648,192 @@ def _gateway(gateway_name=None):
+    local_gw = this_host()
+    disk = request.form.get('disk')
+
+    if request.method == 'PUT':
+
+        if disk not in config.config['disks']:
+            return jsonify(message="Disk {} is not defined in the configuration".format(disk)), 400
+
+        for iqn, target in config.config['targets'].items():
+            if disk in target['disks']:
+                return jsonify(message="Disk {} cannot be used because it is already mapped on "
+                                       "target {}".format(disk, iqn)), 400
+
+        pool, image_name = disk.split('.')
+        try:
+            rbd_image = RBDDev(image_name, 0, pool)
+            size = rbd_image.current_size
+            logger.debug("{} size is {}".format(disk, size))
+        except rbd.ImageNotFound:
+            return jsonify(message="Image {} not found".format(disk)), 400
+
+        owner = LUN.get_owner(config.config['gateways'], target_config['portals'])
+        logger.debug("{} owner will be {}".format(disk, owner))
+
+        api_vars = {
+            'disk': disk,
+            'owner': owner,
+            'allocating_host': local_gw
+        }
+        # process local gateway first
+        gateways.insert(0, local_gw)
+        resp_text, resp_code = call_api(gateways, '_targetlun',
+                                        '{}'.format(target_iqn),
+                                        http_method='put',
+                                        api_vars=api_vars)
+        if resp_code != 200:
+            return jsonify(message="Add target LUN mapping failed - "
+                                   "{}".format(resp_text)), resp_code
+
     else:
-        # DELETE gateway request
+        # this is a DELETE request
+
+        if disk not in config.config['disks']:
+            return jsonify(message="Disk {} is not defined in the "
+                                   "configuration".format(disk)), 400
+
+        if disk not in target_config['disks']:
+            return jsonify(message="Disk {} is not defined in target "
+                                   "{}".format(disk, target_iqn)), 400
+
+        for group_name, group in target_config['groups'].items():
+            if disk in group['disks']:
+                return jsonify(message="Disk {} belongs to group "
+                                       "{}".format(disk, group_name)), 400
+
+        api_vars = {
+            'disk': disk,
+            'purge_host': local_gw
+        }
+        # process other gateways first
+        gateways.append(local_gw)
+        resp_text, resp_code = call_api(gateways, '_targetlun',
+                                        '{}'.format(target_iqn),
+                                        http_method='delete',
+                                        api_vars=api_vars)
+        if resp_code != 200:
+            return jsonify(message="Delete target LUN mapping failed - "
+                                   "{}".format(resp_text)), resp_code
+
```

Comment:
I think you need a ```config.refresh()``` here. With a blank configuration, the first time I add a disk it gets sent to the second node. The gwcli queries the config from the local node after adding the target LUN but fails since the config from the local node is missing the ```owner``` key in the disk so the gwcli crashes. I added it to my local install, retested, and it worked.

Reply:
... or perhaps the "config" API entry point should always attempt a refresh before returning the config to ensure it's always up-to-date regardless of where changes were made.

Reply:
Good spotting. I haven't detected this because, on my system, the `config_watcher` updated the config before the gwcli request, so everything worked without the explicit config.refresh.

Should be fixed now, thanks.